proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,532 @@
+`true` to cause all but the latest secret to be revoked.  In the event of a client secret disclosure, a "hard" rotation
+may be performed by setting both `.spec.generateNewSecret` and `.spec.revokeOldSecrets` to `true` (this will revoke all
+pre-existing client secrets and return a newly generated secret).  Leaving both of these fields set to `false` will
+simply return the number of existing client secrets via the `.status.totalClientSecrets` field (this same information is
+available via the `.status.totalClientSecrets` field of the `OIDCClient` resource).
+
+A dynamic client with many client secrets will be slower to authenticate (especially slow when failing to authenticate),
+since each hashed secret must be tried in sequence until one works, and the cost of trying each is relatively high.
+Administrators should take care to remember to come back to use this endpoint again to revoke the old client secrets
+once they are not needed anymore.
+
+An admin would interact with this API by using standard `kubectl` commands:
+
+```yaml
+apiVersion: oauth.virtual.supervisor.pinniped.dev/v1alpha1  # different group to avoid collision with the CRD
+kind: OIDCClientSecretRequest
+metadata:
+  name: client.oauth.pinniped.dev-j77kz
+  namespace: pinniped-supervisor
+spec:
+  generateNewSecret: true
+```
+
+Assuming the above yaml is stored in `file.yaml`, then running:
+
+`kubectl create -f file.yaml`
+
+would cause the server to respond with (note the custom columns in the table output):
+
+```
+NAMESPACE               NAME                                SECRET     TOTAL
+pinniped-supervisor     client.oauth.pinniped.dev-j77kz     12..xz     2
+```
+
+All client secret hashes for an OAuth client will be stored in `pinniped-storage-oidc-client-secret-<metadata.uid>` where
+`metadata.uid` is a hashed version of the UID generated by the Kuberentes API server for the `OIDCClient` custom resource instance that
+represents the given OAuth client.  This secret will have an owner reference back to the `OIDCClient` to guarantee that
+it is automatically garbage collected if the `OIDCClient` is deleted.  As the client secret lookup is UID based, it is
+resilient against any vulnerabilities that arise from the client ID being re-used over time.  Using the UID also
+prevents any issues that could arise from giving the user control over the secret name (i.e. length issues, collision
+issues, validation issues, etc).  Each client will have a 1:1 mapping with a Kubernetes secret - there would always be
+at most one Kubernetes secret per client.  Since the secret name is deterministic based on the client UID, no reference
+field is required on the client.  Kubernetes secrets have a size limit of ~1 MB which is enough to hold many thousands
+of hashes.  This API will enforce a hard limit of 100 secrets per client (having this many client secrets is likely a
```

Comment:
```suggestion
of hashes.  This API will enforce a hard limit of 5 secrets per client (having this many client secrets is likely a
```

This is what I ended up doing in #1235 to avoid super slow authentications.

Reply:
Updated in 7450fb6c8e9937e43fd5f0f95f2caa674413e3eb.